File: plugins/collector/event/apps/bigdata/rabbitmq.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"strings"
)

const rabbitmqctl = "/usr/sbin/rabbitmqctl"

type Rabbitmq struct {
	version string
}

func (Rabbitmq) Name() string { return "rabbitmq" }

func (Rabbitmq) Type() string { return "bigdata" }

func (r Rabbitmq) Version() string { return r.version }

// run within container or as a service
func (r *Rabbitmq) Match(p *process.Process) bool {
	return p.Name == "beam.smp"
}

func (r *Rabbitmq) Run(p *process.Process) (m map[string]string, err error) {
	// Two situation here:
	// 1. install from yum/apt, deamon process is systemd (ppid 1)
	// 2. pull from the docker, deamon process is rabbitmq-server
	// But it is not an absolutely thing. We need to double check for this
	r.version = ""
	if strings.Contains(p.PpidArgv, "rabbitmq-server") {
		var proc *process.Process
		var fds []string
		proc, err = process.GetProcessInfo(p.PPID, false)
		if err != nil {
			return
		}
		if fds, err = proc.Fds(); err != nil {
			return
		}
		for _, fd := range fds {
			if filepath.Base(fd) != "rabbitmq-server" {
				continue
			}
			// get the sbin
			result, err := apps.ExecuteWithName(p, filepath.Join(filepath.Dir(fd), "rabbitmqctl"), "version")
			if err == nil {
				r.version = strings.TrimSpace(result)
			}
			break
		}
	} else {
		// systemd, try to get the path of rabbitmq
		result, err := apps.ExecuteWithName(p, "rabbitmqctl", "version")
		if err == nil {
			r.version = strings.TrimSpace(result)
		}
	}
	return
}

func init() { apps.Regist(&Rabbitmq{}) }

File: plugins/collector/src/event/apps/bigdata/rabbitmq.rs
use crate::event::apps::{execute_with_name, AppProc, IApp};
use std::collections::HashMap;
use std::path::Path;

pub struct Rabbitmq {
    version: String,
}

impl Rabbitmq {
    pub fn new() -> Self { Rabbitmq { version: String::new() } }
}

impl IApp for Rabbitmq {
    fn name(&self)     -> &'static str { "rabbitmq" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "beam.smp" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        let bin = if p.argv.contains("rabbitmq-server") {
            find_rabbitmqctl_in_fds(p.pid).unwrap_or_else(|| "rabbitmqctl".to_owned())
        } else {
            "rabbitmqctl".to_owned()
        };
        if let Ok(out) = execute_with_name(p, &bin, &["version"]) {
            self.version = out.trim().to_owned();
        }
        Ok(HashMap::new())
    }
}

/// Locate `rabbitmqctl` by finding `rabbitmq-server` in the process's fd links.
fn find_rabbitmqctl_in_fds(pid: i32) -> Option<String> {
    let fd_dir = format!("/proc/{}/fd", pid);
    if let Ok(entries) = std::fs::read_dir(&fd_dir) {
        for entry in entries.flatten() {
            if let Ok(target) = std::fs::read_link(entry.path()) {
                let p = Path::new(&target);
                if p.file_name().map_or(false, |n| n == "rabbitmq-server") {
                    if let Some(dir) = p.parent() {
                        return Some(dir.join("rabbitmqctl").to_string_lossy().into_owned());
                    }
                }
            }
        }
    }
    None
}
